rfc/rfc-56/rfc-56.md

```diff
@@ -0,0 +1,226 @@
+
+Apache Hudi follows the traditional Hive storage layout while writing files on storage:
+- Partitioned Tables: The files are distributed across multiple physical partition folders, under the table's base path.
+- Non Partitioned Tables: The files are stored directly under the table's base path.
+
+While this storage layout scales well for HDFS, it increases the probability of hitting request throttle limits when
+working with cloud object stores like Amazon S3 and others. This is because Amazon S3 and other cloud stores [throttle
+requests based on object prefix](https://aws.amazon.com/premiumsupport/knowledge-center/s3-request-limit-avoid-throttling/).
+Amazon S3 does scale based on request patterns for different prefixes and adds internal partitions (with their own request limits),
+but there can be a 30 - 60 minute wait time before new partitions are created. Thus, all files/objects stored under the
+same table path prefix could result in these request limits being hit for the table prefix, specially as workloads
+scale, and there are several thousands of files being written/updated concurrently. This hurts performance due to
+re-trying of failed requests affecting throughput, and result in occasional failures if the retries are not able to
+succeed either and continue to be throttled.
+
+The traditional storage layout also tightly couples the partitions as folders under the table path. However,
+some users want flexibility to be able to distribute files/partitions under multiple different paths across cloud stores,
+hdfs etc. based on their specific needs. For example, customers have use cases to distribute files for each partition under
+a separate S3 bucket with its individual encryption key. It is not possible to implement such use-cases with Hudi currently.
+
+The high level proposal here is to introduce a new storage layout strategy, where all files are distributed evenly across
+multiple randomly generated prefixes under the Amazon S3 bucket, instead of being stored under a common table path/prefix.
+This would help distribute the requests evenly across different prefixes, resulting in Amazon S3 to create partitions for
+the prefixes each with its own request limit. This significantly reduces the possibility of hitting the request limit
+for a specific prefix/partition.
+
+In addition, we want to expose an interface that provides users the flexibility to implement their own strategy for
+distributing files if using the traditional Hive storage layout or federated storage layer (proposed in this RFC) does
+not meet their use-case.
+
+## Design
+
+### Interface
+
+```java
+/**
+ * Interface for providing storage file locations.
+ */
+public interface FederatedStorageStrategy extends Serializable {
+  /**
+   * Return a fully-qualified storage file location for the given filename.
+   *
+   * @param fileName data file name
+   * @return a fully-qualified location URI for a data file
+   */
+  String storageLocation(String fileName);
```

Comment:
Make API generic to return HudiPhysicalLocation - encapsulate all details to reach a file in cloud?

Reply:
Maybe I'm missing something, but is there anything other than physical path that Hudi needs to reach a file in cloud?

The return value here should be stored in metadata table so Hudi knows where to read it later. If we do need an object here then I think we can extend metadata payload

Reply:
I have a concern similar to @prasannarajaperumal: if we're talking about the URI here we need to clearly define the encoding for such URI. Having something like Hadoop's `Path` or Java's `URI` would have been more clear in terms of expectations. 

WDYT?

Reply:
On the writing side, I think FS is already auto-configured in classes like `HoodieWriteHandle`. Having an object here that wraps `Path` would work but is an additional layer.

I haven't checked what's on the reading side yet but do you see an outstanding blocker of using plain string?